### ISSUE TITLE
IstioOperator spec.meshConfig is not applied

### ISSUE THREAD

**Issue by rocketspacer** [istio/istio#22347]
**Bug description**
**IstioOperator** `spec.meshConfig` is not applied.

**Expected behavior**

When I apply the **IstioOperator** CRD like this
```yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: istio
spec:
  namespace: istio-system
  profile: default
  meshConfig:
    outboundTrafficPolicy:
      mode: REGISTRY_ONLY
```
The generated config map should have synchronized meshConfig

**Steps to reproduce the bug**

1. Run `istioctl manfiest generate -f operator.yaml` with the above manifest
2. See that the output config map still has `outboundTrafficPolicy.mode` set to **ALWAYS_ALLOW**
![image](https://user-images.githubusercontent.com/17026876/77149191-018b4b00-6acc-11ea-838c-f4fd2d4bfabd.png)

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
![image](https://user-images.githubusercontent.com/17026876/77234196-743d1900-6be7-11ea-99d4-3cac6fb97954.png)
![image](https://user-images.githubusercontent.com/17026876/77234210-87e87f80-6be7-11ea-83d7-9f08e4fd193b.png)

**How was Istio installed?**
Using the istioctl

**Environment where bug was observed (cloud vendor, OS, etc)**
Cloud vendor: AWS (EKS)
OS: AWS Container Optimized EC2 Linux

**Comment by clsacramento** [comment]
Hi, 

I managed to get that working with:
spec:
  profile: demo
  values:
    global:
      outboundTrafficPolicy:
        mode: REGISTRY_ONLY


However it is not clear to me why this config does not work under spec.meshConfig...

**Comment by howardjohn** [comment]
I think its specifically this setting, or more likely enums in general. On master:
```
$ cat <<EOF | go run ./istioctl/cmd/istioctl manifest generate -f - | grep outboundTrafficPolicy -C 2
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    outboundTrafficPolicy:
      mode: "ALLOW_ANY"
EOF
Error: error unmarshaling JSON: unknown value "ALLOW_ANY" for enum istio.mesh.v1alpha1.MeshConfig_OutboundTrafficPolicy_Mode:

YAML:
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    outboundTrafficPolicy:
      mode: "ALLOW_ANY"

exit status 1
```

**Comment by howardjohn** [comment]
drain duration is broken as well (I assume all durations)

**Comment by howardjohn** [comment]
https://github.com/istio/istio/pull/22593 may give some more hints about what is going on as well

**Comment by howardjohn** [comment]
Root cause is https://github.com/istio/istio/blob/master/operator/pkg/translate/translate.go#L256 maybe?

**Comment by marshallford** [comment]
I'm also bumping into this error. Specifically when defining a field under `ingressGateways[*].k8s.service`.

**Comment by ostromart** [comment]
That's a pretty good guess. Some places in the code are using regular yaml unmarshal which has problems with enums I believe.

**Comment by richardwxn** [comment]
Yep it is the marshal/unmarshal issue again, we actually had these settings defined in values proto as well, but we use jsonpb to unmarshall for them, which works well with enum.

**Comment by richardwxn** [comment]
I don't want to change back to jsonpb unmarshaller because we intended to replace that with normal yaml unmarshaller before? Another workaround is to patch the meshConfig part of the generated pb.go file to update the json tag, any idea? @ostromart @morvencao

**Comment by morvencao** [comment]
Have tried with following IOP instance:
```
# cat << EOF | out/linux_amd64/istioctl manifest generate -o istio-manifests -f -
> apiVersion: install.istio.io/v1alpha1
> kind: IstioOperator
> spec:
>   meshConfig:
>     outboundTrafficPolicy:
>       mode: 1
> EOF
2020-04-03T03:20:26.722482Z	warn	Config not found: /root/.kube/config
Component dependencies tree:
Base
  Pilot

Rendering manifests to output dir istio-manifests
Rendering: Base
Writing manifest to istio-manifests/Base/Base.yaml
Rendering: Pilot
Writing manifest to istio-manifests/Base/Pilot/Pilot.yaml
Rendering: Cni
Writing manifest to istio-manifests/Base/Pilot/Cni/Cni.yaml
Rendering: IngressGateways
Writing manifest to istio-manifests/Base/Pilot/IngressGateways/IngressGateways.yaml
Rendering: EgressGateways
Writing manifest to istio-manifests/Base/Pilot/EgressGateways/EgressGateways.yaml
Rendering: AddonComponents
Writing manifest to istio-manifests/Base/Pilot/AddonComponents/AddonComponents.yaml
Rendering: Policy
Writing manifest to istio-manifests/Base/Pilot/Policy/Policy.yaml
Rendering: Telemetry
Writing manifest to istio-manifests/Base/Pilot/Telemetry/Telemetry.yaml
Rendering: Galley
Writing manifest to istio-manifests/Base/Pilot/Galley/Galley.yaml
```

Looks like the issue is similar to https://github.com/istio/istio/issues/21554, as @richardwxn mentioned, 
it is caused by the json unmarshaler.

Actually I tried to raised a PR to fix this: https://github.com/istio/api/pull/1363 and tried locally it still doesn't work unfortunately.

**Comment by howardjohn** [comment]
I think its not just the jsonpb vs json, its also because we do not properly distinguish between explicitly setting an option, and it being the default option. Its made worse from the fact that there is about 8 ways to configure every option

**Comment by ostromart** [comment]
Regarding the enum unmarshal problem, I think the issue is that afaics there's no UnmarshalJSON generated for mesh. There is for operator and I think the enums in there do unmarshal correctly with json as a result.

**Comment by morvencao** [comment]
I also thought the root cause is that there's no UnmarshalJSON generated for meshConfig, so I tried to add it in https://github.com/istio/api/pull/1363 
Unfortunately, it was still not working as expected.

**Comment by howardjohn** [comment]
My understanding for registry_only is at some point we lose the information because we cannot distinguish between unset and registry_only since it's the default value

**Comment by morvencao** [comment]
yes, probably.
It's weird because we encountered the same issue for IOP status: https://github.com/istio/api/blob/master/operator/v1alpha1/operator.pb.go#L26-L56
And it was fixed by adding MarshalJSON/UnmarshalJSON.

**Comment by howardjohn** [comment]
It doesn't matter how it is unmarshalled. The problem is once we go from json -> struct we lose information, regardless of how we do the conversion

**Comment by morvencao** [comment]
Agreed.

**Comment by howardjohn** [comment]
https://github.com/istio/istio/pull/22951 should fix this

**Comment by howardjohn** [comment]
Fixed by ^

**Comment by marshallford** [comment]
@howardjohn I see this missed 1.5.2. Will we not see the fix until 1.6?
